pkg/workflows/ring/plugin.go

```diff
@@ -0,0 +1,290 @@
+	return nil
+}
+
+func (p *Plugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *Plugin) Outcome(_ context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	// Load prior state
+	prior := &pb.Outcome{}
+	if outctx.PreviousOutcome == nil {
+		prior.Routes = map[string]*pb.WorkflowRoute{}
+		prior.State = &pb.RoutingState{Id: outctx.SeqNr, State: &pb.RoutingState_RoutableShards{RoutableShards: p.minShardCount}}
+	} else if err := proto.Unmarshal(outctx.PreviousOutcome, prior); err != nil {
+		return nil, err
+	}
+
+	currentShardHealth := make(map[uint32]int)
+	totalObservations := len(aos)
+	allWorkflows := make([]string, 0)
+	nows := make([]time.Time, 0)
+
+	// Collect shard health observations and workflows
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			p.lggr.Warnf("failed to unmarshal observation: %v", err)
+			continue
+		}
+
+		for shardID, healthy := range observation.ShardHealthStatus {
+			if healthy {
+				currentShardHealth[shardID]++
+			}
+		}
+
+		// Collect workflow IDs
+		allWorkflows = append(allWorkflows, observation.WorkflowIDs...)
+
+		// Collect timestamps
+		if observation.Now != nil {
+			nows = append(nows, observation.Now.AsTime())
+		}
+	}
+
+	// Calculate median time
+	now := time.Now()
+	if len(nows) > 0 {
+		slices.SortFunc(nows, time.Time.Compare)
+		now = nows[len(nows)/2]
+	}
+
+	// Deduplicate workflows
+	workflowMap := make(map[string]bool)
+	for _, wf := range allWorkflows {
+		workflowMap[wf] = true
+	}
+	allWorkflows = make([]string, 0, len(workflowMap))
+	for wf := range workflowMap {
+		allWorkflows = append(allWorkflows, wf)
+	}
+	slices.Sort(allWorkflows) // Ensure deterministic order
+
+	// Determine desired shard count based on observations
+	healthyShardCount := uint32(0)
+	for shardID, count := range currentShardHealth {
+		if count > p.config.F {
+			healthyShardCount++
+			// Update store with healthy shard
+			p.store.SetShardHealth(shardID, true)
+		}
+	}
+
+	// Ensure within bounds
+	if healthyShardCount < p.minShardCount {
+		healthyShardCount = p.minShardCount
+	}
+	if healthyShardCount > p.maxShardCount {
+		healthyShardCount = p.maxShardCount
+	}
+
+	// Calculate next state using state machine
+	nextState, err := p.calculateNextState(prior.State, healthyShardCount, now)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use deterministic hashing to assign workflows to shards
+	routes := make(map[string]*pb.WorkflowRoute)
+	for _, wfID := range allWorkflows {
+		assignedShard := p.store.GetShardForWorkflow(wfID)
```

Comment:
We shouldn't be using the local store object here. Outcome needs to be a pure function, otherwise we are risking protocol failures if we happen to have inconsistent store state across nodes.